### PR TITLE
Invoice rate stats improvements

### PR DESCRIPTION
**Pull request by SKollerud**

### CODE REVIEW

**Comment by SKollerud** [comment]
Implementation of #636

**Review by palbock — DISMISSED** [review]

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs [hunk 1]
```diff
@@ -9,5 +9,22 @@ public class InvoiceStatisticsDto
     public decimal[] VacationHours { get; set; }
     public decimal[] InvoiceRate { get; set; }
     public decimal[] NonBillableInvoiceRate { get; set; }
-    public DateTime[] Labels { get; set; }
+    public DateTime[] Start { get; set; }
+    public DateTime[] End { get; set; }
+    
+    public enum InvoicePeriods
+    {
+        Daily,
```
Comment (klyngen) [inline packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs]: Could you add values to these? `Daily = 0`. I know that Enums automatically can be implicitly casted to `int` but that is not obvious to most javascript-monkeys  :monkey:

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs [hunk 2]
```diff
@@ -9,5 +9,22 @@ public class InvoiceStatisticsDto
     public decimal[] VacationHours { get; set; }
     public decimal[] InvoiceRate { get; set; }
     public decimal[] NonBillableInvoiceRate { get; set; }
-    public DateTime[] Labels { get; set; }
+    public DateTime[] Start { get; set; }
```
Comment (klyngen) [inline packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs:12]: These are cool but what would be really smart is if you in addition to these values added the Labels-array.  I don't totally remember the best way of handling it. Maybe adding something like this to bridge the gap:

```csharp
[Obsolete("This attribute will be removed in the next iteration of invoice-rate functionality. Use Start or End instead")]
public DateTime[] Labels { get { return this.Start; } }


```

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceRateService.cs [hunk 3]
```diff
@@ -55,111 +56,145 @@ public async Task<decimal> GetEmployeeInvoiceRateForPeriod(DateTime fromDate, Da
...
+                vacationHours.Add(0);
+                invoiceRates.Add(0);
+                nonBillableInvoiceRate.Add(0);
+                starts.Add(start);
+                ends.Add(end);
+
+                previousPeriodEnd = end;
+            }
+
+            var bh = grouping.Where(timeEntry => GetTaskType(timeEntry) == TaskType.BILLABLE).Sum(timeEntry => timeEntry.Value);
```
Comment (klyngen) [inline packages/api/AlvTime.Business/InvoiceRate/InvoiceRateService.cs]: Consider writing `bh`, `nbh` and `vh` to something easier to read like `billableHours`, `nonBillableHours` and `vacationHours`. #selfDocumentingCode

File: packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs [hunk 4]
```diff
@@ -38,8 +41,8 @@ public async Task<decimal> FetchUserInvoiceRate(DateTime? fromDate, DateTime? to
 
     [HttpGet("InvoiceStatistics")]
     [Authorize(Policy = "AllowPersonalAccessToken")]
-    public async Task<InvoiceStatisticsDto> FetchUserInvoiceStatistics(DateTime fromDate, DateTime toDate)
+    public async Task<InvoiceStatisticsDto> FetchUserInvoiceStatistics(DateTime fromDate, DateTime toDate, InvoicePeriods period, ExtendPeriod extendPeriod, bool includeZeroPeriods)
```
Comment (klyngen) [inline packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs]: Adding a default parameter to InvoicePeriods, making it `InvoicePeriods period = InvoicePeriod.Month` and add the Labels-thing commented earlier and you will have a backend that does not break with any existing functionality :bulb:

File: packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs [hunk 5]
```diff
@@ -1,9 +1,12 @@
 using System;
+using System.Collections.Generic;
```
Comment (klyngen) [inline packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs]: Unused import

File: packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs [hunk 6]
```diff
@@ -1,9 +1,12 @@
 using System;
+using System.Collections.Generic;
 using System.Threading.Tasks;
 using AlvTime.Business.InvoiceRate;
 using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Cors.Infrastructure;
```
Comment (klyngen) [inline packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs]: Unused import

**Review by klyngen — CHANGES_REQUESTED** [review]
I see a lot of cool functionality. I have some nitpicky comments and two suggestions to make the frontend run again. All without touching the frontend :+1:

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceRateService.cs (hunk 3 above)
Comment (SKollerud) [reply packages/api/AlvTime.Business/InvoiceRate/InvoiceRateService.cs]: billableHours, nonBillableHours and vacationHours is already in use, so I need some alternative words. Do you have any suggestions. currentBillableHours, tempBillableHours, billableHours1?

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs (hunk 2 above)
Comment (SKollerud) [reply packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs:12]: Will do!

File: packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs (hunk 1 above)
Comment (SKollerud) [reply packages/api/AlvTime.Business/InvoiceRate/InvoiceStatisticsDto.cs]: Will do!

File: packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs (hunk 4 above)
Comment (SKollerud) [reply packages/api/AlvTimeWebApi/Controllers/InvoiceRateController.cs]: How about setting InvoicePeriod.Month to 0 in the enum instead?

**Comment by SKollerud** [comment]
Included the feedback from high elf Klingen to make js happy.

**Review by klyngen — DISMISSED** [review]
I pulled and built the most recent version of this branch. Works with no changes done to the frontend :+1: 
image

**Review by klyngen — APPROVED** [review]

**Pull request merged**